### PR TITLE
Implement SetItemObserver for observing mutations on a set 

### PR DESCRIPTION
**Pull request by kitchoi**
This PR implements item 10 of #977 for observing mutations on a set (strictly speaking, TraitSet).

- Add `SetChangeEvent` to represent mutations on a set
- Implement `IObserver` interface to
     - Return the TraitSet object for adding/removing notifiers
     - Yield the content of the set for the next observer(s) in the observer graph.
      - Define the notifier such that `SetChangeEvent` will be created for the user's change handler.
      - Implement logic for maintaining downstream observers
- Implement the interface of `IObservable` for `TraitSet`. With that, any subclass of `TraitSet` can be used with SetItemObserver.

**Checklist**
- [x] Tests
- [x] Update API reference (`docs/source/traits_api_reference`)
- ~Update User manual (`docs/source/traits_user_manual`)~
- ~Update type annotation hints in `traits-stubs`~

### CODE REVIEW

File: traits/observers/_set_item_observer.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        ----------
+        graph : ObserverGraph
+            The graph this observer is part of.
+
+        Yields
+        ------
+        ObserverGraph
+        """
+        # Unlike CTrait, no need to handle trait_added
+        yield from ()
```
Comment (mdickinson) [inline traits/observers/_set_item_observer.py:167]: That's a neat trick for an empty generator. I've always used horrible things like `if False: yield None`, but this is better. (It's always felt like a violation of EIBTI that a Python generator function is _implicitly_ a generator function because it uses `yield` somewhere, rather than because it's declared to be a generator function somehow.)

File: traits/observers/tests/test_set_change_event.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            notifiers=[notifier],
+        )
+
+        # when
+        trait_set.add(4)
+
+        # then
+        event, = events
+        self.assertEqual(event.added, set([4]))
+        self.assertEqual(event.removed, set([]))
```
Comment (mdickinson) [inline traits/observers/tests/test_set_change_event.py:57]: Nitpick: `set()` is the usual way to spell the empty set. (Could also use `{4}` in the line above.)

File: traits/observers/tests/test_set_item_observer.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    DummyObservable,
+    DummyObserver,
+    DummyNotifier,
+)
+from traits.trait_set_object import TraitSet
+from traits.trait_types import Set
+
+
+def create_observer(**kwargs):
+    """ Convenient function for creating SetItemObserver with default values.
```
Comment (mdickinson) [inline traits/observers/tests/test_set_item_observer.py:28]: typo: "Convenient" -> "Convenience"

File: traits/observers/tests/test_set_item_observer.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    """
+    values = dict(
+        notify=True,
+        optional=False,
+    )
+    values.update(kwargs)
```
Comment (mdickinson) [inline traits/observers/tests/test_set_item_observer.py:34]: Yet another place where Python 3.9's `|` operator would have been handy ...  (Nothing actionable here; I just wish that the `|` operator for `dict`s had been introduced long ago.)

**Comment by kitchoi** [comment]
> Should we wait until #1071 is merged before merging this one, just in case there are last minute changes to #1071 that require corresponding for-consistency changes in this PR?

Yes!  Will do that.

**Comment by mdickinson** [comment]
One typo to fix and merge conflicts to sort out; other than that, I think this is ready to merge.

File: traits/trait_set_object.py [hunk 5]
```diff
@@ -15,6 +15,7 @@
 
 from traits.trait_base import _validate_everything
 from traits.trait_errors import TraitError
+from traits.observers._i_observable import IObservable as _IObservable
```
Comment (kitchoi) [inline traits/trait_set_object.py:18]: Import order... and I want to get rid this preceding underscore. Will do that in the separate PR that does the same thing to all the other observables.

Comment (kitchoi) [reply traits/trait_set_object.py:18]: https://github.com/enthought/traits/pull/1085
